plutus-core/testlib/PlutusCore/Generators/PIR/Substitutions.hs

```diff
@@ -0,0 +1,312 @@
+variable). Type unification could take a substitution as an input an extend it instead of always
+making the initial substitution empty, but we didn't need this extra functionality, so we didn't
+implement it for the sake of keeping the interface to the unifier simpler.
+
+Type unification proceeds by recursion on the spines of the two types. We don't attempt to do
+higher-order unification at all, so we don't have the usual unification business like pruning,
+blocking, pattern unification etc. In particular, in our setting two type applications can unify
+only if the functions and the arguments unify separately (with higher-order unification it would be
+possible to unify type applications whose function parts differ, because functions compute and the
+results may unify even if the functions don't).
+
+We handle 'TyIFix' and 'TyFun' the same way as type applications by decomposing them into their
+constituents and unifying them separately. These two would look the same in a higher-order setting.
+
+'TyLam' and 'TyForall' being binders require a bit of special care: we rename the bound variables in
+the bodies of the binders, so that the two sides of the unification equation bind the same variable,
+see Note [Renaming during unification]. We do not attempt to perform unification modulo eta.
+
+Two 'TyBuiltin's unify when they are the same built-in type.
+
+The only remaining case is when one of the sides is a type variable (or both). There's a number of
+checks that we perform in order to ensure that it's fine to resolve the variable as the given type,
+those are documented alongside the code performing the checks.
+-}
+
+{- Note [Renaming during unification]
+When we go under twin binders, we rename the variables that the binders introduce, so that they have
+the same name. This requires travering both the sides of the unification problem before unification
+can proceed. This is inefficient, instead we could do the same thing that we do for checking
+equality and keep two separate maps tracking renamings of local variables. Although the types AST is
+lazy, so maybe it's not a big deal in the end. Although even with laziness, we still have quadratic
+behavior currently.
+
+Note that generating a new name is also a rather expensive operation requiring to traverse the whole
+context, so that's another source of inefficiency.
+-}
+
+-- See Note [The unification algorithm].
+-- | Perform unification. Sound but not complete.
+unifyType :: TypeCtx
+          -- ^ Type context
+          -> Set TyName
+          -- ^ @flex@, the flexible variables (those that can be unified)
+          -> Type TyName DefaultUni ()
+          -- ^ @t1@
+          -> Type TyName DefaultUni ()
+          -- ^ @t2@
+          -> Either String TypeSub
+          -- ^ Either an error or a substitution (from a subset of @flex@) unifying @t1@ and @t2@
+unifyType ctx flex a0 b0 =
+    execStateT (runReaderT (goType (normalizeTy a0) (normalizeTy b0)) Set.empty) Map.empty
```

Comment:
Because type checking is modulo normalization of types. Hence type unification should also be modulo normalization of types. It's the usual unification business, you can see some examples inline. E.g. we could solve `_f x =?= x -> x` as `_f := \x -> x -> x`.

Reply:
Right, but what happens if we don't normalize? I guess we just won't notice that some things unify? It sounds like we had some kind of unification of applications, but we probably want to get rid of the ones that actually compute? Can we end up with new computing-applications as we progress through unification? I don't really know what I'm talking about, I should actually read up on unification...

Reply:
> I guess we just won't notice that some things unify?

Yes.

> It sounds like we had some kind of unification of applications, but we probably want to get rid of the ones that actually compute?

Yes, we do absolutely want that. The more complicated your unification machinery is, the more you want to simplify the equation before moving to solving it, particularly since with a complicated unification algorithm you're going to be looking a lot for certain patterns like a metavariable applied to a bunch of variables (a.k.a. Miller pattern unification) etc. I can't imagine any reason to start unification on non-normalized types.

> Can we end up with new computing-applications as we progress through unification?

In the current implementation -- no. In theory -- yes, but it's really complicated.